clone_on_ref_ptr lint is not always trivial to resolve

<!--
Hi there! Whether you've come to make a suggestion for a new lint, an improvement to an existing lint or to report a bug or a false positive in Clippy, you've come to the right place.

If you want to report that Clippy does not compile, please be sure to be using the *latest version* of *Rust nightly*! Compiler plugins are highly unstable and will only work with a nightly Rust for now. If you are but still have a problem, please let us now!

Thank you for using Clippy!

Write your comment below this line: -->

We've just spotted the [`clone_on_ref_ptr`](https://github.com/rust-lang-nursery/rust-clippy/issues/1645) lint throw up clippy warnings in our code. In an effort to go through and fix them, we realized that is isn't actually _always_ very trivial to do so.

I've had a go at condensing our code down into a minimal repro. It has something to with type inference, and trait objects (e.g. interpreting an `Arc` of a `struct` as an `Arc` of a `trait`).

### Concrete example: [playground](https://play.rust-lang.org/?gist=34968767c572c840f2d14e759f00512f&version=stable)

`let _z = Bar(y.clone());` fails the clippy lint: 

```
warning: using '.clone()' on a ref-counted pointer
  --> src/main.rs:13:18
   |
13 |     let _z = Bar(y.clone());
   |                  ^^^^^^^^^ help: try this: `Arc::clone(&y)`
   |
   = note: #[warn(clone_on_ref_ptr)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.160/index.html#clone_on_ref_ptr
```

**The suggested fix, however, does not compile.**

```
error[E0308]: mismatched types
  --> src/main.rs:13:29
   |
13 |     let _z = Bar(Arc::clone(&y));
   |                             ^^ expected trait Foo, found struct `Bar`
   |
   = note: expected type `&std::sync::Arc<Foo>`
              found type `&std::sync::Arc<Bar>`
```

One way to actually get this to work, is to provide explicit types (`Arc::<Bar>::clone`). However, in real life the struct `Bar` might actually have a bunch of generics, which are incredibly tedious to write out (even with underscores) - for example `Arc::Bar<_, _, _, _>::clone`.

### Why I think this is an issue

In this particular case, the clippy warning is trying to get us to write the clone in a slightly different way, to make it clear that we're cloning an `Arc` (cheap). However, that slightly different way doesn't actually compile. This forces us to disable the warning, or jump through several hoops to get some quite inelegant code which satisfies the compiler and clippy.

--- a/clippy_lints/src/methods.rs
+++ b/clippy_lints/src/methods.rs
@@ -361,9 +361,8 @@ declare_lint! {
 /// ```rust
 /// x.clone()
 /// ```
-declare_lint! {
+declare_restriction_lint! {
     pub CLONE_ON_REF_PTR,
-    Warn,
     "using 'clone' on a ref-counted pointer"
 }
 
@@ -1013,24 +1012,26 @@ fn lint_clone_on_copy(cx: &LateContext, expr: &hir::Expr, arg: &hir::Expr, arg_t
 fn lint_clone_on_ref_ptr(cx: &LateContext, expr: &hir::Expr, arg: &hir::Expr) {
     let (obj_ty, _) = walk_ptrs_ty_depth(cx.tables.expr_ty(arg));
 
-    let caller_type = if match_type(cx, obj_ty, &paths::RC) {
-        "Rc"
-    } else if match_type(cx, obj_ty, &paths::ARC) {
-        "Arc"
-    } else if match_type(cx, obj_ty, &paths::WEAK_RC) || match_type(cx, obj_ty, &paths::WEAK_ARC) {
-        "Weak"
-    } else {
-        return;
-    };
+    if let ty::TyAdt(_, subst) = obj_ty.sty {
+        let caller_type = if match_type(cx, obj_ty, &paths::RC) {
+            "Rc"
+        } else if match_type(cx, obj_ty, &paths::ARC) {
+            "Arc"
+        } else if match_type(cx, obj_ty, &paths::WEAK_RC) || match_type(cx, obj_ty, &paths::WEAK_ARC) {
+            "Weak"
+        } else {
+            return;
+        };
 
-    span_lint_and_sugg(
-        cx,
-        CLONE_ON_REF_PTR,
-        expr.span,
-        "using '.clone()' on a ref-counted pointer",
-        "try this",
-        format!("{}::clone(&{})", caller_type, snippet(cx, arg.span, "_")),
-    );
+        span_lint_and_sugg(
+            cx,
+            CLONE_ON_REF_PTR,
+            expr.span,
+            "using '.clone()' on a ref-counted pointer",
+            "try this",
+            format!("{}<{}>::clone(&{})", caller_type, subst.type_at(0), snippet(cx, arg.span, "_")),
+        );
+    }
 }
 
 

--- a/clippy_lints/src/methods.rs
+++ b/clippy_lints/src/methods.rs
@@ -1029,7 +1029,7 @@ fn lint_clone_on_ref_ptr(cx: &LateContext, expr: &hir::Expr, arg: &hir::Expr) {
             expr.span,
             "using '.clone()' on a ref-counted pointer",
             "try this",
-            format!("{}<{}>::clone(&{})", caller_type, subst.type_at(0), snippet(cx, arg.span, "_")),
+            format!("{}::<{}>::clone(&{})", caller_type, subst.type_at(0), snippet(cx, arg.span, "_")),
         );
     }
 }

--- a/tests/ui/unnecessary_clone.rs
+++ b/tests/ui/unnecessary_clone.rs
@@ -1,9 +1,14 @@
+#![warn(clone_on_ref_ptr)]
 #![allow(unused)]
 
 use std::collections::HashSet;
 use std::collections::VecDeque;
 use std::rc::{self, Rc};
 use std::sync::{self, Arc};
+
+trait SomeTrait {}
+struct SomeImpl;
+impl SomeTrait for SomeImpl {}
 
 fn main() {}
 
@@ -34,7 +39,8 @@ fn clone_on_ref_ptr() {
     arc_weak.clone();
     sync::Weak::clone(&arc_weak);
 
-
+    let x = Arc::new(SomeImpl);
+    let _: Arc<SomeTrait> = x.clone(); 
 }
 
 fn clone_on_copy_generic<T: Copy>(t: T) {

--- a/tests/ui/unnecessary_clone.stderr
+++ b/tests/ui/unnecessary_clone.stderr
@@ -16,7 +16,7 @@ error: using '.clone()' on a ref-counted pointer
   --> $DIR/unnecessary_clone.rs:30:5
    |
 30 |     rc.clone();
-   |     ^^^^^^^^^^ help: try this: `Rc<bool>::clone(&rc)`
+   |     ^^^^^^^^^^ help: try this: `Rc::<bool>::clone(&rc)`
    |
    = note: `-D clone-on-ref-ptr` implied by `-D warnings`
 
@@ -24,25 +24,25 @@ error: using '.clone()' on a ref-counted pointer
   --> $DIR/unnecessary_clone.rs:33:5
    |
 33 |     arc.clone();
-   |     ^^^^^^^^^^^ help: try this: `Arc<bool>::clone(&arc)`
+   |     ^^^^^^^^^^^ help: try this: `Arc::<bool>::clone(&arc)`
 
 error: using '.clone()' on a ref-counted pointer
   --> $DIR/unnecessary_clone.rs:36:5
    |
 36 |     rcweak.clone();
-   |     ^^^^^^^^^^^^^^ help: try this: `Weak<bool>::clone(&rcweak)`
+   |     ^^^^^^^^^^^^^^ help: try this: `Weak::<bool>::clone(&rcweak)`
 
 error: using '.clone()' on a ref-counted pointer
   --> $DIR/unnecessary_clone.rs:39:5
    |
 39 |     arc_weak.clone();
-   |     ^^^^^^^^^^^^^^^^ help: try this: `Weak<bool>::clone(&arc_weak)`
+   |     ^^^^^^^^^^^^^^^^ help: try this: `Weak::<bool>::clone(&arc_weak)`
 
 error: using '.clone()' on a ref-counted pointer
   --> $DIR/unnecessary_clone.rs:43:29
    |
 43 |     let _: Arc<SomeTrait> = x.clone(); 
-   |                             ^^^^^^^^^ help: try this: `Arc<SomeImpl>::clone(&x)`
+   |                             ^^^^^^^^^ help: try this: `Arc::<SomeImpl>::clone(&x)`
 
 error: using `clone` on a `Copy` type
   --> $DIR/unnecessary_clone.rs:47:5

--- a/tests/ui/unnecessary_clone.stderr
+++ b/tests/ui/unnecessary_clone.stderr
@@ -1,75 +1,81 @@
 error: using `clone` on a `Copy` type
-  --> $DIR/unnecessary_clone.rs:11:5
+  --> $DIR/unnecessary_clone.rs:16:5
    |
-11 |     42.clone();
+16 |     42.clone();
    |     ^^^^^^^^^^ help: try removing the `clone` call: `42`
    |
    = note: `-D clone-on-copy` implied by `-D warnings`
 
 error: using `clone` on a `Copy` type
-  --> $DIR/unnecessary_clone.rs:15:5
+  --> $DIR/unnecessary_clone.rs:20:5
    |
-15 |     (&42).clone();
+20 |     (&42).clone();
    |     ^^^^^^^^^^^^^ help: try dereferencing it: `*(&42)`
 
 error: using '.clone()' on a ref-counted pointer
-  --> $DIR/unnecessary_clone.rs:25:5
+  --> $DIR/unnecessary_clone.rs:30:5
    |
-25 |     rc.clone();
-   |     ^^^^^^^^^^ help: try this: `Rc::clone(&rc)`
+30 |     rc.clone();
+   |     ^^^^^^^^^^ help: try this: `Rc<bool>::clone(&rc)`
    |
    = note: `-D clone-on-ref-ptr` implied by `-D warnings`
 
 error: using '.clone()' on a ref-counted pointer
-  --> $DIR/unnecessary_clone.rs:28:5
+  --> $DIR/unnecessary_clone.rs:33:5
    |
-28 |     arc.clone();
-   |     ^^^^^^^^^^^ help: try this: `Arc::clone(&arc)`
+33 |     arc.clone();
+   |     ^^^^^^^^^^^ help: try this: `Arc<bool>::clone(&arc)`
 
 error: using '.clone()' on a ref-counted pointer
-  --> $DIR/unnecessary_clone.rs:31:5
+  --> $DIR/unnecessary_clone.rs:36:5
    |
-31 |     rcweak.clone();
-   |     ^^^^^^^^^^^^^^ help: try this: `Weak::clone(&rcweak)`
+36 |     rcweak.clone();
+   |     ^^^^^^^^^^^^^^ help: try this: `Weak<bool>::clone(&rcweak)`
 
 error: using '.clone()' on a ref-counted pointer
-  --> $DIR/unnecessary_clone.rs:34:5
+  --> $DIR/unnecessary_clone.rs:39:5
    |
-34 |     arc_weak.clone();
-   |     ^^^^^^^^^^^^^^^^ help: try this: `Weak::clone(&arc_weak)`
+39 |     arc_weak.clone();
+   |     ^^^^^^^^^^^^^^^^ help: try this: `Weak<bool>::clone(&arc_weak)`
+
+error: using '.clone()' on a ref-counted pointer
+  --> $DIR/unnecessary_clone.rs:43:29
+   |
+43 |     let _: Arc<SomeTrait> = x.clone(); 
+   |                             ^^^^^^^^^ help: try this: `Arc<SomeImpl>::clone(&x)`
 
 error: using `clone` on a `Copy` type
-  --> $DIR/unnecessary_clone.rs:41:5
+  --> $DIR/unnecessary_clone.rs:47:5
    |
-41 |     t.clone();
+47 |     t.clone();
    |     ^^^^^^^^^ help: try removing the `clone` call: `t`
 
 error: using `clone` on a `Copy` type
-  --> $DIR/unnecessary_clone.rs:43:5
+  --> $DIR/unnecessary_clone.rs:49:5
    |
-43 |     Some(t).clone();
+49 |     Some(t).clone();
    |     ^^^^^^^^^^^^^^^ help: try removing the `clone` call: `Some(t)`
 
 error: using `clone` on a double-reference; this will copy the reference instead of cloning the inner type
-  --> $DIR/unnecessary_clone.rs:49:22
+  --> $DIR/unnecessary_clone.rs:55:22
    |
-49 |     let z: &Vec<_> = y.clone();
+55 |     let z: &Vec<_> = y.clone();
    |                      ^^^^^^^^^
    |
    = note: `-D clone-double-ref` implied by `-D warnings`
 help: try dereferencing it
    |
-49 |     let z: &Vec<_> = &(*y).clone();
+55 |     let z: &Vec<_> = &(*y).clone();
    |                      ^^^^^^^^^^^^^
 help: or try being explicit about what type to clone
    |
-49 |     let z: &Vec<_> = &std::vec::Vec<i32>::clone(y);
+55 |     let z: &Vec<_> = &std::vec::Vec<i32>::clone(y);
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: called `cloned().collect()` on a slice to create a `Vec`. Calling `to_vec()` is both faster and more readable
-  --> $DIR/unnecessary_clone.rs:56:27
+  --> $DIR/unnecessary_clone.rs:62:27
    |
-56 |     let v2 : Vec<isize> = v.iter().cloned().collect();
+62 |     let v2 : Vec<isize> = v.iter().cloned().collect();
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D iter-cloned-collect` implied by `-D warnings`

